Added doctest unit test framework

RevoluPowered opened the pull request:
This uses doctest to implement unit testing for the core of godot.

![UnitTesting Prototype](https://user-images.githubusercontent.com/748770/86599545-ac017e00-bf96-11ea-96f3-c3c1a0ec4c00.gif)


**This includes the following:**
- a very quick test runner of a single test by running the godot executable.
- the third party library with license file (MIT) and appropriate documentation in the third party folder.
- modifications to SCSub file to make tests part of main executable as linking tests statically can be hard (need a second opinion on this approach too)

**Benefits**
- horrible regressions which we don't want to come back can have unit tests!
- Simple to use (https://github.com/godotengine/godot/pull/40148/files#diff-b3c825b8305cec3713d66729d1efcd90)
- No boilerplate
- headless or non headless possibility
- CI integration (https://github.com/onqtam/doctest/blob/master/doc/markdown/reporters.md)
  - has all the weird and wonderful reporters for various CI tools
  - CI can be run with headless tests like for string, basis and math checks we do not require a renderer.
- Less files
- follows KISS principle and supports a lot of other useful features https://blog.jetbrains.com/rscpp/better-ways-testing-with-doctest/
- less code per test (because you don't need to copy paste a test harness)

Current test runner can be executed by running ./bin/godot.x11.tools.64 --test string, or by specific test name

Demo with ```./godot --test``` argument

![image](https://user-images.githubusercontent.com/748770/67679796-cdc75180-f981-11e9-9355-e742808aa350.png)

Demo with backward compatibility with other core tests ```--test math```
![image](https://user-images.githubusercontent.com/748770/67679871-fea78680-f981-11e9-829e-87543e3403fd.png)

Articles of supporting info:
- https://github.com/onqtam/doctest/blob/master/doc/markdown/benchmarks.md
- https://blog.jetbrains.com/rscpp/better-ways-testing-with-doctest/

RevoluPowered commented:
#30743 - updated doctest for 4.0 fixed conflicts

akien-mga commented:
> Current test runner can be executed by running ./bin/godot.x11.tools.64 --test doctest

How would future tests be handled? Would it be `--test doctest` as an entry point to run all tests, or can we keep using `--test string` instead, as it's not particularly relevant what framework is used when calling the test?

Once this is merged, I think we should port all other tests to doctest and document the workflow.

akien-mga commented:
Some style issues: https://travis-ci.org/github/godotengine/godot/jobs/705243011

Also make sure to use braces around single-line `if` or `for` blocks.

You lost some of @aaronfranke's changes in `test_string` while rebasing, so it fails building on Clang: https://travis-ci.org/github/godotengine/godot/jobs/705243015
(Be sure to keep the MinGW-specific hack too.)

RevoluPowered commented:
@akien-mga I have applied the requested changes, but made a slight improvement to the mingw specific section.

I used PRId64 to correct this issue as it handles the platform specific issue here for us, but I need to test this on mingw, so will do this in a bit and if it works would it be okay to keep this?

The fix is based on https://stackoverflow.com/questions/9225567/how-to-print-a-int64-t-type-in-c this article, just so we don't have to duplicate code or do extra casting

RevoluPowered commented:
> > Current test runner can be executed by running ./bin/godot.x11.tools.64 --test doctest
> 
> How would future tests be handled? Would it be `--test doctest` as an entry point to run all tests, or can we keep using `--test string` instead, as it's not particularly relevant what framework is used when calling the test?
> 
> Once this is merged, I think we should port all other tests to doctest and document the workflow.

I will write a quick fix for this now, to make it so

RevoluPowered commented:
OK all points raised should now be sorted, I expanded the run functionality to make it more usable for regression testing too, it's super simple lets you run whatever specific test you would like too, also leaves legacy compatibility in.

demo gif
![UnitTesting Prototype](https://user-images.githubusercontent.com/748770/86599545-ac017e00-bf96-11ea-96f3-c3c1a0ec4c00.gif)

@aaronfranke can you possibly take a look at the fix I applied here, I think it could be simpler for godot to use this instead, it should remove the need for the preprocessor.

https://github.com/godotengine/godot/pull/40148/commits/1cd932db36a31c1464222bf15086ea9638750040

reference: https://stackoverflow.com/questions/9225567/how-to-print-a-int64-t-type-in-c

akien-mga commented on main/tests/test_string.h:
```diff
@@ -31,12 +31,1009 @@
 #ifndef TEST_STRING_H
 #define TEST_STRING_H
 
+#include <thirdparty/doctest/doctest.h>
```
`< >` is for system headers, so this should use `" "`.

I'd put it further down in its own section before the system includes.

akien-mga commented on main/tests/test_string.h:
```diff
@@ -31,12 +31,1009 @@
+#include "core/os/os.h"
 #include "core/ustring.h"
 
+#include "modules/regex/regex.h"
```
This should be conditional to the regex module being enabled, as it was previously.

RevoluPowered replied:
Fixed

akien-mga commented on main/tests/test_validate_testing.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef TEST_VALIDATE_TESTING_H
+#define TEST_VALIDATE_TESTING_H
+
+#include "core/os/os.h"
+#include <thirdparty/doctest/doctest.h>
```
```suggestion
#include "core/os/os.h"

#include "thirdparty/doctest/doctest.h"
```

akien-mga commented on thirdparty/README.md:
```diff
@@ -74,6 +74,14 @@ Files extracted from upstream source:
 - all .cpp, .h, and .txt files in ConvectionKernels/
 
 
+## doctest
+- Upstream: https://github.com/onqtam/doctest.git
+- Version: 2ce8c59 (2.3.3)
+- License: MIT License
```
```suggestion
## doctest

- Upstream: https://github.com/onqtam/doctest
- Version: 2.3.3 (2019)
- License: MIT
```

akien-mga commented on thirdparty/README.md:
```diff
@@ -74,6 +74,14 @@ Files extracted from upstream source:
 - all .cpp, .h, and .txt files in ConvectionKernels/
 
 
+## doctest
```
Also add to `COPYRIGHT.txt`:
```diff
diff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index fc3079c361..ec67193e13 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -126,6 +126,11 @@ Copyright: 2018, Eric Lasota
  2018, Microsoft Corp.
 License: Expat
 
+Files: ./thirdparty/doctest/
+Comment: doctest
+Copyright: 2019, Viktor Kirilov
+License: Expat
+
 Files: ./thirdparty/enet/
 Comment: ENet
 Copyright: 2002-2020, Lee Salzman
```

RevoluPowered replied:
Added will commit soon.

akien-mga commented on thirdparty/doctest/doctest.h:
```diff
@@ -0,0 +1,6205 @@
+#define DOCTEST_VERSION_MAJOR 2
+#define DOCTEST_VERSION_MINOR 4
+#define DOCTEST_VERSION_PATCH 0
+#define DOCTEST_VERSION_STR "2.4.0"
```
You documented version 2.3.3 in `thirdparty/README.md`.

RevoluPowered replied:
will update this now

aaronfranke replied:
Why start at 2.4.0?

RevoluPowered replied:
Corrected

akien-mga commented on main/SCsub:
```diff
@@ -8,6 +8,7 @@ import main_builders
 env.main_sources = []
 
 env.add_source_files(env.main_sources, "*.cpp")
+env.add_source_files(env.main_sources, "tests/*.cpp")
```
They're already compiled by `main/tests/SCsub`.

akien-mga commented on main/SCsub:
```diff
@@ -23,6 +24,5 @@ env.CommandNoCache("#main/app_icon.gen.h", "#main/app_icon.png", run_in_subproce
 
 if env["tools"]:
     SConscript("tests/SCsub")
-
```
I liked this line.

akien-mga commented on SConstruct:
```diff
@@ -641,6 +641,9 @@ if selected_platform in platform_list:
         )
 
+    # enable test framework globally and inform it of configuration method
+    env.Append(CPPDEFINES=["DOCTEST_CONFIG_IMPLEMENT"])
```
Can we have this in `main/tests/SCsub`? Not sure why it would have to be in `SConstruct`.

RevoluPowered replied:
It needs to be global for the project so that other parts of the engine can use the tests

RevoluPowered commented:
After some thought, I think we need to approach testing differently a little.

Our test code should run the 'FIXUP' startup logic if required, and the test_main() needs to start a particular thing it should do it.

Every test like string, basis and fbx can run without the engine running, but only runs the engine due to the startup code.

I believe we should make main(argc,argv) run tests and decouple it entirely from the engine startup logic. 

It's going to be much more obvious and allow us to pass arguments in a much simpler way while also giving us the benefit of just running the tests you specify, or all the tests. and also allowing you to supply doctest arguments.

I believe ./godot.linuxbsd.tools --test should override all arguments supplied and pass the rest to doctest's handler.

RevoluPowered commented:
OK so windows main is fixed, and Linux unit tests are fixed, MacOS unit tests also run.

The system now uses a define to 'inject' before Godot's real main logic, but this is the most elegant approach we can have. I have implemented the 'real raw basics' required to initialize the engine singleton OS properly, and this seems to satisfy the engine enough to run tests heedlessly, and the good thing is the CI is running them too, for each editor build it will run 'unit tests' step, I want to have these as a 'stage' under checks, for now though we can't save binaries, so this is the only option right now.

Android and iOS need looked into, as I didn't really have time to do this.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -216,6 +219,12 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      # patch the sources list to use the default ubuntu repository (to prevent azure giving us packages...)
```
Be consistent :)
```suggestion
      # Patch the sources list to use the default ubuntu repository (to prevent azure giving us packages...)
```

akien-mga replied:
Though this is already merged so it should be rebased to remove this code.

akien-mga commented on COPYRIGHT.txt:
```diff
@@ -126,6 +126,11 @@ Copyright: 2018, Eric Lasota
 
+Files: ./thirdparty/doctest/
+Comment: doctest
+Copyright: 2019, Viktor Kirilov
```
2020.

akien-mga replied:
Er no, nevermind. License file states "2016-2019" so that's what we should use: https://github.com/onqtam/doctest/blob/2.4.0/LICENSE.txt
```suggestion
Copyright: 2016-2019, Viktor Kirilov
```

akien-mga commented on main/tests/test_main.cpp:
```diff
@@ -34,23 +34,28 @@
+#include "test_ordered_hash_map.h"
 #include "test_string.h"
+#include "test_basis.h"
+#include "test_validate_testing.h"
+#include "thirdparty/doctest/doctest.h"
```
```suggestion
#include "test_validate_testing.h"

#include "thirdparty/doctest/doctest.h"
```

akien-mga commented on main/tests/test_main.cpp:
```diff
@@ -34,23 +34,28 @@
+#include "test_oa_hash_map.h"
+#include "test_ordered_hash_map.h"
 #include "test_string.h"
+#include "test_basis.h"
```
Wrong order.

akien-mga commented on misc/scripts/sources.list:
```diff
@@ -0,0 +1,10 @@
+deb http://gb.archive.ubuntu.com/ubuntu/ focal main restricted
```
Note: When rebasing, make sure to remove this file (I moved it).

RevoluPowered replied:
ok doing this now

akien-mga commented on thirdparty/README.md:
```diff
@@ -74,6 +74,14 @@ Files extracted from upstream source:
 
+## doctest
+- Upstream: https://github.com/onqtam/doctest.git
+- Version: 1c8da00 (2.4.0)
```
```suggestion
- Version: 2.4.0 (1c8da00c978c19e00a434b2b1f854fcffc9fba35, 2020)
```

akien-mga commented on thirdparty/README.md:
```diff
@@ -74,6 +74,14 @@ Files extracted from upstream source:
+## doctest
+- Upstream: https://github.com/onqtam/doctest.git
+- Version: 1c8da00 (2.4.0)
+- License: MIT License
```
```suggestion
- License: MIT
```

akien-mga commented on thirdparty/README.md:
```diff
@@ -74,6 +74,14 @@ Files extracted from upstream source:
 
 
+## doctest
+- Upstream: https://github.com/onqtam/doctest.git
```
```suggestion
- Upstream: https://github.com/onqtam/doctest
```
I had already asked all those changes :)

akien-mga commented on platform/windows/godot_windows.cpp:
```diff
@@ -145,6 +145,8 @@ int widechar_main(int argc, wchar_t **argv) {
 	for (int i = 0; i < argc; ++i) {
 		argv_utf8[i] = wc_to_utf8(argv[i]);
 	}
+        
```
Trailing whitespace.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -55,7 +55,9 @@ jobs:
       run: |
         scons -j2 verbose=yes warnings=all werror=yes platform=windows tools=yes target=release_debug
-
+    - name: Unit Tests
```
```suggestion
    - name: Unit tests
```

akien-mga replied:
Keep the line.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -163,7 +165,9 @@ jobs:
         run: |
           scons -j2 verbose=yes warnings=all werror=yes platform=osx tools=yes target=release_debug
-
+      - name: Unit Tests
```
And keep that line.

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -206,7 +210,6 @@ jobs:
           SCONS_CACHE: ${{github.workspace}}/.scons_cache/
         run: |
           scons -j2 verbose=yes warnings=all werror=yes platform=osx target=release tools=no
-
```
:angry:

akien-mga commented on .github/workflows/main.yml:
```diff
@@ -259,6 +268,9 @@ jobs:
           SCONS_CACHE: ${{github.workspace}}/.scons_cache/
         run: |
           scons -j2 verbose=yes warnings=all werror=yes platform=linuxbsd tools=yes target=release_debug
+      - name: Unit Tests
```
```suggestion

      - name: Unit tests
```

RevoluPowered committed:
t Add unit testing to Godot using DocTest and added to GitHub Actions CI

Implements exit codes into the engine so tests can return their statuses.
Ideally we don't do this, and we use FIXUP logic to 'begin' and 'end' the engine execution for tests specifically.

Since realistically we're initialising the engine here we don't want to do that, since String should not require an engine startup to test a single header.

This lowers the complexity of running the unit tests and even for
physics should be possible to implement such a fix.

akien-mga commented:
You need to run clang-format (and squash): https://github.com/godotengine/godot/pull/40148/checks?check_run_id=891900765

And also rebase on `master` to get the javascript CI fix.

RevoluPowered commented:
Updated runner to take real arguments, but need to see how portable this solution is

./godot --test --help will return the help for doctest itself now, as I resolved a bug with the wchar_t conversion to char

RevoluPowered commented:
Fixed portability of the strcpy its now strncpy since this is safer

RevoluPowered commented:
Ran some more tests and it seems to be functioning normally now and the platform issues seem to be resolved, formatting should be correct.

akien-mga commented:
Nice work, I'll try to test it today.

@vnen @faless You might want to take a look, and see if you like the new format to write some tests for GDScript and networking respectively.

RevoluPowered commented:
> Nice work, I'll try to test it today.
> 
> @vnen @Faless You might want to take a look, and see if you like the new format to write some tests for GDScript and networking respectively.

I had to remove the legacy tests unfortunately so we need to port them

akien-mga commented:
> I had to remove the legacy tests unfortunately so we need to port them

That's fine, I don't think there was much to salvage in most legacy tests in the first place.

It's better to start cleanly with the new framework, and then port and modernize the tests which were up-to-date enough with the Godot API (like hashmaps and AStar tests).

RevoluPowered commented:
> > 
> 
> That's fine, I don't think there was much to salvage in most legacy tests in the first place.
> 
> It's better to start cleanly with the new framework, and then port and modernize the tests which were up-to-date enough with the Godot API (like hashmaps and AStar tests).

Cool, so I left the real files in main/tests/ but they do nothing at the moment to clarify that I didn't remove the actual files, just the legacy test runner, so whoever ports them its simpler, I will try to help with this effort and get them across but need to do this after merge, basis, fbx and math will probably be the ones I port after merge

RevoluPowered commented:
Possibly we need some system for measuring the code coverage later on too.

akien-mga commented:
Needs a rebase after #40556.

RevoluPowered commented:
> Needs a rebase after #40556.

Done, I ensured no extra whitespace.

akien-mga commented on .github/workflows/linux_builds.yml:
```diff
@@ -62,6 +62,11 @@ jobs:
+      # Execute unit tests for the editor
+      - name: Unit Tests
+        run: |
+          ./bin/godot.linuxbsd.opt.tools.64 --test
```
This now needs to be:
```suggestion
          ./bin/godot.linuxbsd.opt.tools.64.mono --test
```

Only for Linux, Mono builds are not done on other platforms for now.

akien-mga commented on .github/workflows/macos_builds.yml:
```diff
@@ -51,6 +51,11 @@ jobs:
         run: |
           scons -j2 verbose=yes warnings=all werror=yes platform=osx tools=yes target=release_debug
 
+      # execute unit tests for the editor
```
```suggestion
      # Execute unit tests for the editor
```

Consistency! :)

akien-mga commented on .github/workflows/windows_builds.yml:
```diff
@@ -56,6 +56,11 @@ jobs:
       run: |
         scons -j2 verbose=yes warnings=all werror=yes platform=windows tools=yes target=release_debug
 
+    # execute unit tests for the editor
```
```suggestion
    # Execute unit tests for the editor
```

RevoluPowered replied:
fixed!

RevoluPowered commented:
**For Devs:**
To use the unit testing PR you can include the header from thirdparty/doctest/doctest.h it's single header only.

In order to make sure your test is built, once you write your test include the header for it in ```main/tests/test_main.cpp``` this ensures it is built as part of the same static linkage to godot's test harness.

To run tests you can use ```./godot --test``` after running this will mean that you no longer can use godot's arguments, you must now use doctest arguments.

**The basics:**

```cpp
TEST_CASE("some regression test") {
    CHECK(something_worked());
}
```
**Reference guide**
https://github.com/onqtam/doctest

**Things which wont work**
```cpp
// by design errors a check should be a SINGULAR check.
CHECK(something && something_else);
CHECK(something || something_else);
```

akien-mga commented:
Thanks a ton!

aaronfranke commented:
`--test` needs to be added to the `--help` page. Currently ` --help | grep test` returns nothing.

Calinou commented:
@aaronfranke I can see it at the end of the text returned by `--help`:

```text
--test [--help]                  Run unit tests. Use --test --help for more information.
```

Xrayez commented:
Yes, this was done in #40832, the old tests are inaccessible but are still there, but some of those are resurrected now: #41355. See preliminary documentation about tests and those additional test tools in godotengine/godot-docs#4017.
